lib/proxy/robust-sockjs.js

```diff
@@ -198,40 +127,73 @@ function RobustSockJSRegistry(timeout){
       // we're shutting down.
       return robustConn._readyState;
     };
-    this.url = conn.url;
-    this.address = conn.address;
-    this.headers = conn.headers;
     this._withheld = {timer: null, events: []};
 
-    // Override the underlying connection's emit so we can echo from our
-    // RobustConn.
-    conn._oldEmit = conn.emit;
-    conn.emit = function(type) {
-      function doEmit(){
-        // We need to emit on the connections[id] object.
-        robustConn.emit.apply(robustConn, arguments);
-        conn._oldEmit.apply(conn, arguments);
-      }
-      if (type === 'end' || type === 'close'){
-        logger.trace("Withholding event '" + type + "' from robust connection " + id);
-        robustConn._withheld.events.push({arg: arguments});
-        if (!robustConn._withheld.timer){
-          robustConn._withheld.timer = setTimeout(function(){
-            // If time has passed, actually kill the connection by emitting the
-            // withheld events of close and/or end.
-            var evt;
-            while ((evt = robustConn._withheld.events.pop())){
-              doEmit.apply(null, evt.arg);
-            }
-            logger.debug("Closing robust connection " + id);
-            delete self._connections[id];
-            robustConn._readyState = 3; // Mark as closed.
-          }, self._timeout);
+    // Swap out the SockJS connection behind this RobustConn
+    this.set = function(conn){
+      if (robustConn._conn) {
+        // Retire old connection
+
+        // Restore the original emit method so we don't see the bubbled
+        // close/end events on our RobustConn.
+        robustConn._conn.emit = robustConn._conn._oldEmit;
+        // Close the underlying stale SockJS Connection
+        robustConn._conn.close();
+
+        // Clear out any pending close/end messages, as we no longer plan to close.
+        if (robustConn._withheld){
+          robustConn._withheld.events = [];
+          clearTimeout(robustConn._withheld.timer);
+          robustConn._withheld.timer = 0;
+        }
+
+        // Write any buffered events to the new connection
+        var args;
+        while ((args = robustConn._buffer.shift())){
+          conn.write.apply(conn, args.arg);
         }
-      } else {
-        doEmit.apply(null, arguments);
       }
+
+      // Set the new connection
+
+      // Update all of our properties
+      this._conn = conn;
+      this.url = conn.url;
+      this.address = conn.address;
+      this.headers = conn.headers;
+
+      // Override the underlying connection's emit so we can echo from our
+      // RobustConn.
+      conn._oldEmit = conn.emit;
+      conn.emit = function(type) {
```

Comment:
Yeah. I did this rather than having to bring in eventEmitter2 or limit ourselves to some subset of events. But I probably could have safely just picked the ones we know about.